contracts/ERC721MLite.sol

```diff
@@ -0,0 +1,568 @@
+import "@openzeppelin/contracts/utils/cryptography/SignatureChecker.sol";
+import "erc721a/contracts/extensions/ERC721AQueryable.sol";
+import "./IERC721M.sol";
+
+/**
+ * @title ERC721MLite
+ *
+ * @dev ERC721A subclass with MagicEden launchpad features including
+ *  - multiple minting stages with time-based auto stage switch
+ *  - global and stage wallet-level minting limit
+ *  - whitelist using merkle tree
+ *  - crossmint support
+ *  - anti-botting
+ */
+contract ERC721MLite is IERC721M, ERC721AQueryable, Ownable, ReentrancyGuard {
+    using ECDSA for bytes32;
+
+    // Whether this contract is mintable.
+    bool private _mintable;
+
+    // Specify how long a signature from cosigner is valid for, recommend 300 seconds.
+    uint64 private _timestampExpirySeconds;
+
+    // The address of the cosigner server.
+    address private _cosigner;
+
+    // The total mintable supply.
+    uint256 internal _maxMintableSupply;
+
+    // Global wallet limit, across all stages.
+    uint256 private _globalWalletLimit;
+
+    // Current base URI.
+    string private _currentBaseURI;
+
+    // The suffix for the token URL, e.g. ".json".
+    string private _tokenURISuffix;
+
+    // Mint stage infomation. See MintStageInfo for details.
+    MintStageInfo[] private _mintStages;
+
+    // Minted count per stage per wallet.
+    mapping(uint256 => mapping(address => uint32))
+        private _stageMintedCountsPerWallet;
+
+    // Minted count per stage.
+    mapping(uint256 => uint256) private _stageMintedCounts;
+
+    constructor(
+        string memory collectionName,
+        string memory collectionSymbol,
+        string memory tokenURISuffix,
+        uint256 maxMintableSupply,
+        uint256 globalWalletLimit,
+        address cosigner,
+        uint64 timestampExpirySeconds
+    ) ERC721A(collectionName, collectionSymbol) {
+        if (globalWalletLimit > maxMintableSupply)
+            revert GlobalWalletLimitOverflow();
+        _mintable = false;
+        _maxMintableSupply = maxMintableSupply;
+        _globalWalletLimit = globalWalletLimit;
+        _tokenURISuffix = tokenURISuffix;
+        _cosigner = cosigner; // ethers.constants.AddressZero for no cosigning
+        _timestampExpirySeconds = timestampExpirySeconds;
+    }
+
+    /**
+     * @dev Returns whether mintable.
+     */
+    modifier canMint() {
+        if (!_mintable) revert NotMintable();
+        _;
+    }
+
+    /**
+     * @dev Returns whether NOT mintable.
+     */
+    modifier cannotMint() {
+        if (_mintable) revert Mintable();
+        _;
+    }
+
+    /**
+     * @dev Returns whether it has enough supply for the given qty.
+     */
+    modifier hasSupply(uint256 qty) {
+        if (totalSupply() + qty > _maxMintableSupply) revert NoSupplyLeft();
+        _;
+    }
+
+    /**
+     * @dev Returns crossmint address. Always returns 0x0.
+     */
+    function getCrossmintAddress() external pure override returns (address) {
```

Comment:
Would it make sense to just remove this function?

Reply:
This is defined in IERC721M so I would prefer to refactor it later (don't bundle changes of ERC721M )

Reply:
There are also some functions having no references. Plan to clean them separately.